btree/btree.c

```diff
@@ -3460,21 +3519,236 @@ static int64_t btree_get_tick(struct m0_sm_op *smop)
-	switch(bop->bo_op.o_sm.sm_state)
-	{
-		case P_INIT:
-			return P_ACT;
+	struct m0_btree_op    *bop   = M0_AMB(bop, smop, bo_op);
+	struct td             *tree  = bop->bo_arbor->t_desc;
+	struct m0_btree_oimpl *oi    = bop->bo_i;
+	struct level          *lev;
 
-		case P_ACT:
-			return P_DONE;
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+		if ((bop->bo_flags & BOF_COOKIE) &&
+		    cookie_is_set(&bop->bo_rec.r_key.k_cookie))
+			return P_COOKIE;
+		else
+			return P_SETUP;
+	case P_COOKIE:
+		if (cookie_is_valid(tree, &bop->bo_rec.r_key.k_cookie))
+			return P_LOCK;
+		else
+			return P_SETUP;
+	case P_SETUP:
+		bop->bo_arbor->t_height = tree->t_height;
+		oi = level_alloc(tree->t_height);
+		if (oi == NULL)
+			return fail(bop, M0_ERR(-ENOMEM));
+		return P_LOCKALL;
+	case P_LOCKALL:
+		if (bop->bo_flags & BOF_LOCKALL)
+			return m0_sm_op_sub(&bop->bo_op, P_LOCK, P_DOWN);
+		/** Fall through if LOCKALL flag is not set. */
+	case P_DOWN:
+		oi->i_used  = 0;
+		oi->i_pivot = -1;
+		return node_get(&oi->i_nop, tree, &tree->t_root->n_addr,
+				P_NEXTDOWN);
+	case P_NEXTDOWN:
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			struct slot    s = {};
+			struct segaddr child;
 
-		default:
-			return 0;
+			lev = &oi->i_level[oi->i_used];
+			lev->l_node = oi->i_nop.no_node;
+			s.s_node = oi->i_nop.no_node;
+			lev->l_seq = lev->l_node->n_seq;
+			oi->i_key_found = node_find(&s, &bop->bo_rec.r_key);
+			lev->l_idx = s.s_idx;
+
+			if (node_level(s.s_node) > 0) {
+				if (oi->i_key_found) {
+					s.s_idx++;
+					lev->l_idx++;
+				}
+				/**
+				 * Check if the node has valid left or right
+				 * index based on next previous flag. If valid
+				 * left/right index found. mark this level as
+				 * pivot level.
+				 */
+				if (((bop->bo_flags & BOF_NEXT) &&
+				    (lev->l_idx < node_count(lev->l_node))) ||
+				    ((bop->bo_flags & BOF_PREV) &&
+				    (lev->l_idx > 0)))
+					oi->i_pivot = oi->i_used;
+
+				node_child(&s, &child);
+				if (!address_in_segment(child)) {
+					node_op_fini(&oi->i_nop);
+					return fail(bop, M0_ERR(-EFAULT));
+				}
+				oi->i_used++;
+				return node_get(&oi->i_nop, tree,
+						&child, P_NEXTDOWN);
+			} else	{
+				/* Get sibling index based on PREV/NEXT flag. */
+				lev->l_idx = sibling_index_get(s.s_idx,
+							       bop->bo_flags,
+							       oi->i_key_found);
+				/**
+				 * In the following cases jump to LOCK state:
+				 * 1. Normal pass without sibling load flag.
+				 * 2. Sibling load flag is set and the found key
+				 *    idx is withing the valid index range of
+				 *    the node.
+				 * 3.i_pivot is equal to -1. It means, tree
+				 *   traversal reached at the leaf level without
+				 *   finding any valid sibling in the non-levels.
+				 *   This indicates that the search key is the
+				 *   boundary key (rightmost for NEXT flag and
+				 *   leftmost for PREV flag).
+				 */
+				if (!(bop->bo_flags & BOF_SIBLING) ||
+				    (index_is_valid(lev)) ||
```

Comment:
here first you can check if index is valid , if it is not valid and oi->i_pivot !=-1, then you might want to set the sibling flag and load the sibling, you can avoid traversing of tree again from P_ACT.

Reply:
Yes, I had thought of it earlier but was afraid of checking this in unlock state. I will make the changes as per suggestion and see if it there is any issue. Thanks.

Reply:
Even if we are doing this in unlock mode, in path check we will check the node validity and sequence counter of both the nodes, current node and sibling node, so if there is any change path_check will catch it , so we will not get unexpected result